### ISSUE TITLE
question on parallel forward

### ISSUE THREAD

**Issue by evgenii-egorov** [google-deepmind/synjax#4]
Hello,
This is question, not issue.

I was curios on your implementation of parallel scan for HMM with jax. For my best understanding, example of this can be found [linear_chain_crf.py#L116](https://github.com/google-deepmind/synjax/blob/916fc7217f0df7003ef4edfc8045d8d1ad290d7f/synjax/_src/linear_chain_crf.py#L116)

However, I don't understand what makes execution to force "run in parallel"? Thank you

**Comment by stanojevic** [comment]
Figure 4 in [Rush 2020](https://arxiv.org/pdf/2002.00876.pdf) shows a high level view of how the parallel algorithm works by merging transition matrices for each position in a form of a tree structure.
The for loop [here](https://github.com/google-deepmind/synjax/blob/916fc7217f0df7003ef4edfc8045d8d1ad290d7f/synjax/_src/linear_chain_crf.py#L163) loops over log(n) of tree levels and each tree level's nodes are processed with a single semiring einsum [here](https://github.com/google-deepmind/synjax/blob/916fc7217f0df7003ef4edfc8045d8d1ad290d7f/synjax/_src/linear_chain_crf.py#L160) by merging every even-position node with it's immediate neighbour to the right using a semiring matmul.
After log(n) einsums there will be only one node remaining in the end -- the root node of the tree -- out of which log-partition can be extracted by summing out the remaining axes.
This can be implemented more easily using [jax.lax.associative_scan](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.associative_scan.html). I forgot why I decided not to use jax.lax.associative_scan but the idea of that combinator is the same -- any sequence of associative operations can be computed in parallel by forming a tree structure of computation. This usually involves large temporary tensors so it will probably be slow unless there are custom kernels to optimize memory access.

**Comment by evgenii-egorov** [comment]
Thank you! I understand structure of algorithm, so for me a problem is how to be parallel at each level of tree.

So parallelisation of processing einsums on each level in your case is due to merging them to one einsum call. This is use the fact that transition tensor has the same shape for each tilmestep, if I understand correctly? 

I saw jax.lax.associative_scan, however their implementation is based on recursion. Hence, I guess the length of compiling will grow with the number of time steps (in HMM case).

Thank you for answer

**Comment by stanojevic** [comment]
Yes, parallelisation is due to a single einsum op for each level of the tree.

It is true that compilation time for jax.lax.associative_scan depends on the input size, but that also applies to this implementation in SynJax because loop is explicitly unrolled. I don't think that's a big problem in most situations because log(n) grows very slowly and if the n is really very large it's more likely that memory will be a bigger issue than the compilation time.

If the compilation time is the main problem, it could be solved by replacing mentioned for loop with jax.lax.scan. Because jax.lax.scan requires looping with same shapes, the 'def loop' function will need to be rewritten to always return consistently same shapes and use mask for controlling which entries are active. It would make implementation more cluttered without a clear speed benefit.
